source/language/openpulse.rst

```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-   defcal rz(angle[20]:theta) %q {
-     frameof("drive", %q).phase -= theta;
+.. code-block:: javascript
+
+   defcal prelude {
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+    }
+
+   defcal access_frames $q {
+     frame driveframe2 = copy(driveframe1); // `driveframe1` globally scoped
    }
 
-Here's an example qubit spectroscopy experiment.
 
-.. code-block:: none
+To allow for parameterized access to these globally defined frames, we can also define a ``map``.
+
+.. code-block:: javascript
 
-   qubit q;
+   defcal prelude {
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+     ...
+     map[int, frame] frame_mapping = {0: driveframe0, 1: driveframe1};
+    }
 
-   const shots = 1000;
-   const start = 5e9; // 5 GHz
-   const end = 6e9; // 6 GHz
-   const points = 50;
+   defcal access_frames $q {
+     frame driveframe2 = copy(driveframe0); // `driveframe0` globally scoped
+     frame driveframe3 = frame_mapping[$q]; // access `driveframe0` or `driveframe1` by the integer label
+   }
 
-   complex[32] iq, average;
-   complex[32] output[points];
 
-   const driveframe = frameof("drive", q);
+If there are multiple frames associated with a qubit, the map can be extended using a
+``list``
 
-   for p in [0 : points-1] {
-     // The key line: pick the frequency to sample
-     driveframe.frequency = start + (end-start) * p / points;
-     output[p] = 0;
+.. code-block:: javascript
 
-     for i in [0 : shots-1] {
-       // Assumes suitable calibrations for reset, x, and measure_iq
-       // and that the x gate references the same "drive" frame
-       reset q;
-       x q;
-       measure_iq q -> iq;
+   defcal prelude {
+     frame frame_01 = Frame(1.0, 5e9, 0.0);
+     frame frame_02 = Frame(1.0, 6e9, 0.0);
+     ...
+     map[int, list[frame]] frame_mapping = {0: [frame_01, frame_02], ...};
+    }
 
-       output[p] = (output[p] * i + iq) / (i + 1);
-     }
+   defcal access_frames $q {
+     frame qframe_01 = frame_mapping[$q][0];
+     frame qframe_12 = frame_mapping[$q][1];
+   }
+
+Note that ``map`` and ``list`` are simple base containers and are not tied to the ``frame``
+construction, they  simply make it easier to parametrize the access to these frames and could
+be useful in mapping and/or collecting other OpenQASM types.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   driveframe.phase += pi/4;
```

Comment:
What update time does this operation take place at?

Reply:
Since each `frame` is responsible for keeping its own time, the update will occur at the current time of the `driveframe`.